### DISCUSSION TITLE
Talk = GPT-2 + Whisper + WASM

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/whisper.cpp#167] [Show and tell]
I just had an awesome idea:

Make a web-page that:
- Listens when someone speaks
- Transcribes the words using [WASM Whisper](https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.wasm)
- Generates a new sentence using [WASM GPT-2](https://github.com/ggerganov/ggml/tree/master/examples/gpt-2)
- Uses [Web Speech API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API) to synthesise the speech and play it on the speakers.

**All of this running locally in the browser - no server required**

I have all the ingredients and I think the performance is just enough. I just have to put it together.
The total data that the page will have to load on startup (probably using Fetch API) is:
- 74 MB for the Whisper `tiny.en` model
- 240 MB for the GPT-2 `small` model
- Web Speech API is built-in in modern browsers

I think it will be very fun because you could talk to the web-page or even add extra devices that talk to each other only through the mic and the speakers. For example, you simply open the page on your phone and tablet and put them next to each other - listen to them talk about something 😄 

Any ideas to make this even more fun?

---

**Update:**

This is now fully functional at: https://whisper.ggerganov.com/talk/

Source code is here: https://github.com/ggerganov/whisper.cpp/tree/master/examples/talk.wasm

Looking for beta testers, feedback and ideas for improvement!

https://user-images.githubusercontent.com/1991296/203411580-fedb4839-05e4-4474-8364-aaf1e9a9b615.mp4

**Comment by eschmidbauer** [comment]
this sounds really fun!

**Comment by ggerganov** [comment]
So.. this is turning out to be even better than I expected 😆 

https://user-images.githubusercontent.com/1991296/202914175-115793b1-d32e-4aaa-a45b-59e313707ff6.mp4

**Comment by Vuizur** [comment]
These results are extremely impressive! I recently tried to implement something [similar](https://github.com/Vuizur/gpt3-chatbot) in Python, only not locally, but instead using different online APIs, but it felt worse than your demo video because Whisper is much better than the free Google Speech Recognition API (and your optimized version runs significantly better on CPU than the standard Whisper Python lib I tried) :).

**Comment by ggerganov** [comment]
And here is a less cringe video to demonstrate the capabilities of this implementation:

https://user-images.githubusercontent.com/1991296/203845553-f7b44e13-9a15-4fc8-b518-ae8f4c6770fe.mp4

These are 2 Chrome tabs talking and being nice to each other using the microphone and the speakers of a Macbook.

**Comment by beartell** [comment]
There is an error on browser Firefox and Chrome..

![resim](https://user-images.githubusercontent.com/870265/203992168-66a54598-9547-479e-8c7f-ac50adbcb412.png)

**Reply by ggerganov** [reply]
Likely, you haven't enabled cross-origin isolation on your HTTP server.
For more information, see my https://github.com/ggerganov/whisper.cpp/issues/88#issuecomment-1290976962

**Comment by beartell** [comment]
Amazing solution. Works like a charm ;)

Pazar, Kasım 27, 2022 13:17 +03 tarihinde, Georgi Gerganov ***@***.***> şunu yazdı:
 
 
Likely, you haven't enabled cross-origin isolation on your HTTP server.
For more information, see my #88 (comment)

**Comment by aklos** [comment]
This is great, been trying to make the same thing except through terminal. Why GPT-2 instead of the GPT-3 text-davinci-003 model?

**Reply by ghost** [reply]
This is good, but if you integrate whisper with chatgpt, it'd be even more awesome.

**Reply by ggerganov** [reply]
To use GPT-3, you have to make API requests and pay for that. With GPT-2, it runs ***on your computer*** for free.

**Reply by ghost** [reply]
@ggerganov I meant the recently released one, I think it's free, I've seen some scripts and trying those. I mean, the newer chatbot is pretty advanced compared to the dull GPT-2. Very stark difference.

**Reply by aklos** [reply]
@kingofdelphi I'm pretty sure OpenAI might ban people who use workarounds to use the ChatGPT API. They haven't released a public API yet, so anything you do for that right now would be probably violating ToS.

@ggerganov To be fair though, the API requests are cheap enough that I still prefer to use GPT-3 for personal apps. It might be a good idea to allow people to switch to GPT-3 (using their own API key) if they want to.

**Reply by Anx2k** [reply]
I wonder if running something like GPT-JT would be possible:
https://www.together.xyz/blog/releasing-v1-of-gpt-jt-powered-by-open-source-ai
Really the GPT-2 part is the only kind of disappointing aspect, as Whisper is great and even the speech synth is more than acceptable, but GPT-2 is giving me Eliza flashbacks.  ;)

**Reply by ggerganov** [reply]
You won't be able to load GPT-JT in a web-page - it's too big. You can easily run `talk` in the command-line using GPT-J or GPT-JT, but then you won't have speech synthesiser. But I'm pretty sure somebody will hack together something very soon.

**Reply by ggerganov** [reply]
So I made a terminal version running the GPT-2 1558M XL model. Unfortunately, the conversations are still not very coherent. I think the prompt strategy has to be improved somehow, but I cannot figure out how.

https://user-images.githubusercontent.com/1991296/206805012-48e71cc2-588d-4745-8798-c1c70ea3b40d.mp4

Source code: https://github.com/ggerganov/whisper.cpp/blob/master/examples/talk

**Comment by waynenilsen** [comment]
should use hot mic method hold down space to talk so you can take longer pauses while you think about what to say release spacebar to translate speech to text

**Comment by jinsu35** [comment]
I think it would be great to have inference of a text-to-speech model like [this](https://github.com/coqui-ai/TTS).
How would one go about doing this?

**Comment by lili21** [comment]
how about run the wasm on a server?

**Comment by wagler** [comment]
Can't you accelerate the model inference with the GPU via WebGPU for C++?
